bibtex-actions.el

```diff
@@ -234,12 +234,50 @@ If FORCE-REBUILD-CACHE is t, force reloading the cache."
     (bibtex-actions-refresh force-rebuild-cache))
   bibtex-actions--candidates-cache)
 
+(defun bibtex-actions-complete-key-at-point ()
+  "Complete citation key at point.
+
+When inserting '@' in a buffer the capf UI will present user with
+a list of entries, from which they can narrow against title,
+author, etc., and then select one.  This function will then return
+the key 'key', resulting in '@key' at point."
+
+  ;; Activate when user enters '@'.
+  ;;
+  ;; Notes:
+  ;;
+  ;;   - pandoc syntax is [@key], or can also be just @key.
+  ;;   - org syntax is stricter; [cite:see @key1 p23]
+  ;;   - latex uses keys, but sans '@' (I don't want to worry about this ATM though)
+  ;;
+  ;; Maybe this should be constrained further for org?
+  ;;
+  ;; Or, given that pandoc syntax can be used in org context, maybe condition
+  ;; this on 'bibtex-completion-format-citation-functions'?
+  (when (looking-back "@[a-zA-Z0-9]*" 5)
+    (let ((candidates (bibtex-actions--get-candidates)))
+      (list
+       (save-excursion
+         (backward-word)
+         (point))
+       (point)
+       (mapcar #'cdr candidates) :exclusive 'no))))
+
+(defun bibtex-actions-capf ()
+  "My test capf; alt."
+  (when (looking-back "@[a-zA-Z0-9]*" 5)
+    (let* ((candidates (bibtex-actions--get-candidates))
+           ;; FIX
+           (cand (match-string-no-properties 1))
```

Comment:
I need some way to save the selected candidate string, so that I can use it to look up the associated citekey.

Reply:
I'm not sure I understand what you need (or what doesn't) work, but it seems like you should simply call `match-string` directly after `looking-back`. To return what you have found at via `looking-back`, use `(match-string-no-properties 0)`

```emacs-lisp
(when (looking-back regex 5)
  (let* ((match (match-string-no-properties 0))
           (candidates (bibtex-actions--get-candidates))
;;           .....assoc....
   )))
```

Reply:
Right now, I get "no match" when invoking this.

Earlier, I got correct UI (per screenshot), but the returned string is the full candidate, rather than its key, which is stored in the cdr.

In the CR function, we search the candidates list, store the selected ones, and then iterate through those to return the list of keys.

Here's there's just one key though.

So I am looking for something similar.

My problem so far is there's more than one piece that confuses me!

Reply:
Here's the initial CR (before multiple) function, for example:

``` elisp
(defun bibtex-actions--completing-read ()
  "Read bibtex-completion entries for completion using 'completing-read'."
  (bibtex-completion-init)
  (when-let ((candidates (bibtex-actions--get-candidates))
             ;; store selected candidate
             (chosen
              (completing-read
               "BibTeX entries: "
               (lambda (string predicate action)
                 (if (eq action 'metadata)
                     '(metadata
                       ;; (annotation-function . bibtex-completion--annotation)
                       (category . bibtex))
                   (complete-with-action action candidates string predicate))))))
    ;; return selected candidate cdr (key)
    (cdr (assoc chosen candidates))))
```

Reply:
Added `bibtex-actions-insert-key-2" alternative, which was attempting to mirrror the CR function.

"No match."

Reply:
Sorry, I'm out for now, I don't have the time to test it on my machine, and I really don't know this capf stuff...

Reply:
No problem; thanks for the help!